pkg/ipam/allocator.go

```diff
@@ -277,3 +313,72 @@ func (ipam *IPAM) Dump() (allocv4 map[string]string, allocv6 map[string]string,
 
 	return
 }
+
+// StartExpirationTimer installs an expiration timer for a previously allocated
+// IP. Unless StopExpirationTimer is called in time, the IP will be released
+// again after expiration of the specified timeout. The function will return a
+// UUID representing the unique allocation attempt. The same UUID must be
+// passed into StopExpirationTimer again.
+//
+// This function is to be used as allocation and use of an IP can be controlled
+// by an external entity and that external entity can disappear. Therefore such
+// users should register an expiration timer before returning the IP and then
+// stop the expiration timer when the IP has been used.
+func (ipam *IPAM) StartExpirationTimer(ip net.IP, timeout time.Duration) (string, error) {
+	ipam.allocatorMutex.Lock()
+	defer ipam.allocatorMutex.Unlock()
+
+	ipString := ip.String()
+	if _, ok := ipam.expirationTimers[ipString]; ok {
+		return "", fmt.Errorf("expiration timer already registered")
+	}
+
+	allocationUUID := uuid.NewUUID().String()
+	ipam.expirationTimers[ipString] = allocationUUID
+
+	go func(ip net.IP, allocationUUID string, timeout time.Duration) {
+		ipString := ip.String()
+		time.Sleep(timeout)
+
+		ipam.allocatorMutex.Lock()
+		defer ipam.allocatorMutex.Unlock()
+
+		if currentUUID, ok := ipam.expirationTimers[ipString]; ok {
+			if currentUUID == allocationUUID {
+				scopedLog := log.WithFields(logrus.Fields{"ip": ipString, "uuid": allocationUUID})
+				if err := ipam.releaseIPLocked(ip); err != nil {
+					scopedLog.WithError(err).Warning("Unable to release IP after expiration")
+				}
+				scopedLog.Warning("Released IP after expiration")
```

Comment:
line 352 should be in an else, otherwise both Warning messages will be confusing